src/workerd/api/r2-bucket.c++

```diff
@@ -299,7 +299,12 @@ jsg::Promise<kj::Maybe<jsg::Ref<R2Bucket::HeadResult>>> R2Bucket::head(
     auto requestJson = json.encode(requestBuilder);
 
     auto bucket = adminBucket.map([](const auto& b) { return kj::str(b); });
-    auto promise = doR2HTTPGetRequest(kj::mv(client), kj::mv(requestJson), kj::mv(bucket));
+    kj::Vector<kj::StringPtr> path;
+    KJ_IF_MAYBE(b, bucket) {
+      path.add(*b);
+    }
```

Comment:
I think something like

```
kj::StringPtr path[2]
int numComponents = 0
KJ_IF_MAYBE(a, adminAccount) {
  path[numComponents++] = *a;
}
KJ_IF_MAYBE(b, adminBucket) {
  path[numComponents++] = *b;
}

// kj::arrayPtr(path, numComponents) is what you pass into doR2HTTPPutRequests.
```

might be more idiomatic in terms of doing all of this without any memory allocations.

Also not that the duplication of `adminAccount`/`adminBucket` is no longer necessary because that responsibility has been delegated into the internals of `doR2HTTPPutRequest`. In terms of a helper function it could look something like:

```
kj::ArrayPtr<kj::StringPtr> fillPath(kj::StringPtr pathStorage[2], std::initializer_list<const kj::Maybe<kj::String>>& components) {
  int numComponents = 0;
  for (const auto& component: components) {
    KJ_IF_MAYBE(c, component) {
      pathStorage[numComponents++] = *c;
    }
  }
  return kj::arrayPtr(pathStorage, numComponents) ;
}
```

and to use it
```
kj::StringPtr components[2];
auto path = fillPath(components, { adminAccount, adminBucket });
```

I *think* that's directionally how you could do it although there may be syntax errors in the above.

Reply:
Agreed that'd it'd be nice to both reduce the duplication and avoid `kj::Vector`'s allocations. @vlovich's code looks like it'll work, with the caveat that it may want to assert the number of components is never greater than 2.

Reply:
The `fillPath` method you shared @vlovich didn't quite work (`lvalue` vs `rvalue` issue) so I simplified it to be less generic and clever, to make the compiler happy.

Reply:
Yeah. Your way is also cleaner since the storage you're passing in can only ever be 2 anyway